cashu/mint/db/write.py

```diff
@@ -223,3 +229,100 @@ async def _unset_melt_quote_pending(
 
         await self.events.submit(quote_copy)
         return quote_copy
+
+    async def _verify_proofs_and_dlc_registrations(
+        self,
+        registrations: List[Tuple[DiscreetLogContract, DlcFundingProof]],
+    ) -> Tuple[List[Tuple[DiscreetLogContract, DlcFundingProof]], List[DlcFundingProof]]:
+        """
+        Method to check if proofs are already spent or registrations already registered. If they are not, we
+        set them as spent and registered respectively
+        Args:
+            registrations (List[Tuple[DiscreetLogContract, DlcFundingProof]]): List of registrations.
+        Returns:
+            List[Tuple[DiscreetLogContract, DlcFundingProof]]: a list of registered DLCs
+            List[DlcFundingProof]: a list of errors
+        """
+        checked: List[Tuple[DiscreetLogContract, DlcFundingProof]] = []
+        registered: List[Tuple[DiscreetLogContract, DlcFundingProof]] = []
+        errors: List[DlcFundingProof]= []
+        if len(registrations) == 0:
+            logger.trace("Received 0 registrations")
+            return [], []
+        logger.trace("_verify_proofs_and_dlc_registrations acquiring lock")
+        async with self.db.get_connection(lock_table="proofs_used") as conn:
+            for registration in registrations:
+                reg = registration[0]
+                logger.trace("checking whether proofs are already spent")
+                try:
+                    assert reg.inputs
+                    await self.db_read._verify_proofs_spendable(reg.inputs, conn)
+                    await self.db_read._verify_dlc_registrable(reg.dlc_root, conn)
+                    checked.append(registration)
+                except (TokenAlreadySpentError, DlcAlreadyRegisteredError) as e:
+                    logger.trace(f"Proofs already spent for registration {reg.dlc_root}")
+                    errors.append(DlcFundingProof(
+                        dlc_root=reg.dlc_root,
+                        bad_inputs=[DlcBadInput(
+                            index=-1,
```

Comment:
I see the spec is forcing you to put a round peg in a square hole for generic errors. How would you feel about modifying the funding error response objects to support a generic error message for cases like this? 

```diff
{
  "funded": [
    {
      "dlc_root": <hash>,
      "funding_proof": {
        "keyset": <keyset_id_str>,
        "signature": <bip340_sig_hex>
      }
    },
    ...
  ],
  "errors": [
    {
      "dlc_root": <hash>,
      "bad_inputs": [
        {
          "index": <int>,
          "detail": <str>
        },
        ...
      ]
    },
+   {
+     "dlc_root": <hash>,
+     "detail": <str>
+   },
    ...
  ]
}
```

Alternatively if we want to make things simpler, we can remove the `bad_inputs` concept entirely and provide only a `detail` message for the whole DLC funding operation error. This would also make error handling easier, as you wouldn't need to keep track of a list of errors. 

```diff
{
  "funded": [
    {
      "dlc_root": <hash>,
      "funding_proof": {
        "keyset": <keyset_id_str>,
        "signature": <bip340_sig_hex>
      }
    },
    ...
  ],
  "errors": [
    {
      "dlc_root": <hash>,
+     "detail": <str>
-     "bad_inputs": [
-       {
-         "index": <int>,
-         "detail": <str>
-       },
        ...
      ]
    },
    ...
  ]
}
```

Reply:
I think the first solution, with an optional `detail` field in mutual exclusion with `bad_inputs` is the way.